app/controllers/assignments_controller.rb

```diff
@@ -562,8 +583,240 @@ def download_sample_starter_files
+  def config_starter_files(assignment, zip_file)
+    starter_config_file = assignment.is_peer_review? ? :peer_review_starter_files : :starter_files
+    starter_file_settings = build_hash_from_zip(zip_file, starter_config_file).symbolize_keys
+    starter_group_mappings = {}
+    starter_file_settings[:group_information].each do |group|
+      group = group.symbolize_keys
+      file_group = StarterFileGroup.create!(name: group[:name],
+                                            use_rename: group[:use_rename],
+                                            entry_rename: group[:entry_rename],
+                                            assignment: assignment)
+      starter_group_mappings[group[:directory_name]] = file_group
+    end
+    default_name = starter_file_settings[:default_starter_group]
+    if !default_name.nil? && starter_group_mappings.key?(default_name)
+      assignment.default_starter_file_group_id = starter_group_mappings[default_name].id
+    end
+    zip_starter_dir = File.dirname(CONFIG_FILES[starter_config_file])
+    starter_file_glob_pattern = File.join(zip_starter_dir, '**', '*')
+    zip_file.glob(starter_file_glob_pattern) do |entry|
+      next if entry.name == CONFIG_FILES[starter_config_file]
+      # Set working directory to the location of all the starter file content, then find
+      # directory for a starter group and add the file found in that directory to group
+      zip_file_path = Pathname.new(entry.name)
+      starter_base_dir = zip_file_path.relative_path_from(zip_starter_dir)
+      grouping_dir = starter_base_dir.descend.first.to_s
+      starter_file_group = starter_group_mappings[grouping_dir]
+      sub_dir, filename = starter_base_dir.relative_path_from(grouping_dir).split
+      starter_file_dir_path = File.join(starter_file_group.path, sub_dir.to_s)
+      starter_file_name = filename.to_s
+      if entry.directory?
+        FileUtils.mkdir_p(File.join(starter_file_dir_path, starter_file_name))
+      else
+        FileUtils.mkdir_p(starter_file_dir_path)
+        starter_file_content = entry.get_input_stream.read
+        File.write(File.join(starter_file_dir_path, starter_file_name), starter_file_content, mode: 'wb')
+      end
+    end
+    assignment.starter_file_groups.find_each(&:update_entries)
+  end
+
+  # Build the tag/criteria/starter file settings file specified by +hash_to_build+ found in +zip_file+
+  # Delete the file from the +zip_file+ after loading in the content.
+  def build_hash_from_zip(zip_file, hash_to_build)
+    yaml_file = zip_file.get_entry(CONFIG_FILES[hash_to_build])
+    yaml_content = yaml_file.get_input_stream.read.encode(Encoding::UTF_8, 'UTF-8')
+    properties = parse_yaml_content(yaml_content)
+    if [:tags, :peer_review_tags].include?(hash_to_build)
+      properties.each { |row| row[:user] = @current_user.user_name }
+    end
+    properties
+  end
+
+  # Ensure that the +assignment+ type (scanned, timed, neither) matches the params
```

Comment:
This is the only functional part which I'm not a huge fan of. Instead of having the upload link on the assignment form (and passing params for `is_scanned` and `is_timed`), let's move that upload link to the `assignments/index` page's heading links. You can use the name "Upload Configuration Zip File" (from the modal title) to distinguish it from the "Upload" link that already appears. Then, you can remove this method entirely, allowing the instructor to upload any time of assignment by using that modal.

(You can keep the layout changes to the assignment forms that you made, though.)